File: packages/sdk/src/core/utils/validator.ts
import type { EventParams, Result } from '../types';
import { VALIDATION } from '../types';

export class ValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ValidationError';
    Object.setPrototypeOf(this, ValidationError.prototype);
  }
}

function getObjectDepth(
  obj: unknown,
  currentDepth = 1,
  visited = new WeakSet<object>()
): number {
  if (typeof obj !== 'object' || obj === null) {
    return currentDepth;
  }

  if (visited.has(obj)) {
    return currentDepth;
  }
  visited.add(obj);

  const depths: number[] = [];

  if (Array.isArray(obj)) {
    for (const item of obj) {
      depths.push(getObjectDepth(item, currentDepth + 1, visited));
    }
  } else {
    for (const value of Object.values(obj)) {
      depths.push(getObjectDepth(value, currentDepth + 1, visited));
    }
  }

  return depths.length > 0 ? Math.max(...depths) : currentDepth;
}

export function validateEventParams(
  params: EventParams | undefined
): Result<void> {
  if (!params) {
    return { success: true, data: undefined };
  }

  const depth = getObjectDepth(params);
  if (depth > VALIDATION.EVENT_PARAMS.MAX_DEPTH) {
    return {
      success: false,
      error: new ValidationError(
        `Event params exceed maximum depth of ${VALIDATION.EVENT_PARAMS.MAX_DEPTH} (got ${depth})`
      ),
    };
  }

  try {
    const serialized = JSON.stringify(params);
    if (serialized.length > VALIDATION.EVENT_PARAMS.MAX_SIZE) {
      return {
        success: false,
        error: new ValidationError(
          `Event params exceed maximum size of ${VALIDATION.EVENT_PARAMS.MAX_SIZE} bytes`
        ),
      };
    }
  } catch {
    return {
      success: false,
      error: new ValidationError('Event params must be JSON serializable'),
    };
  }

  return { success: true, data: undefined };
}

export function validateEventName(name: string): Result<void> {
  if (name.length < VALIDATION.EVENT_NAME.MIN_LENGTH) {
    return {
      success: false,
      error: new ValidationError(
        `Event name must be at least ${VALIDATION.EVENT_NAME.MIN_LENGTH} character(s)`
      ),
    };
  }

  if (name.length > VALIDATION.EVENT_NAME.MAX_LENGTH) {
    return {
      success: false,
      error: new ValidationError(
        `Event name must not exceed ${VALIDATION.EVENT_NAME.MAX_LENGTH} characters`
      ),
    };
  }

  if (!VALIDATION.EVENT_NAME.PATTERN.test(name)) {
    return {
      success: false,
      error: new ValidationError(
        `Event name must match pattern: ${VALIDATION.EVENT_NAME.PATTERN.toString()}`
      ),
    };
  }

  return { success: true, data: undefined };
}

export function validateDeviceId(deviceId: string): Result<void> {
  if (deviceId.length < VALIDATION.DEVICE_ID.MIN_LENGTH) {
    return {
      success: false,
      error: new ValidationError(
        `Device ID must be at least ${VALIDATION.DEVICE_ID.MIN_LENGTH} characters`
      ),
    };
  }

  if (deviceId.length > VALIDATION.DEVICE_ID.MAX_LENGTH) {
    return {
      success: false,
      error: new ValidationError(
        `Device ID must not exceed ${VALIDATION.DEVICE_ID.MAX_LENGTH} characters`
      ),
    };
  }

  if (!VALIDATION.DEVICE_ID.PATTERN.test(deviceId)) {
    return {
      success: false,
      error: new ValidationError(
        `Device ID must match pattern: ${VALIDATION.DEVICE_ID.PATTERN.toString()}`
      ),
    };
  }

  return { success: true, data: undefined };
}

export function validateSessionId(sessionId: string): Result<void> {
  if (sessionId.length < VALIDATION.SESSION_ID.MIN_LENGTH) {
    return {
      success: false,
      error: new ValidationError(
        `Session ID must be at least ${VALIDATION.SESSION_ID.MIN_LENGTH} characters`
      ),
    };
  }

  if (sessionId.length > VALIDATION.SESSION_ID.MAX_LENGTH) {
    return {
      success: false,
      error: new ValidationError(
        `Session ID must not exceed ${VALIDATION.SESSION_ID.MAX_LENGTH} characters`
      ),
    };
  }

  if (!VALIDATION.SESSION_ID.PATTERN.test(sessionId)) {
    return {
      success: false,
      error: new ValidationError(
        `Session ID must match pattern: ${VALIDATION.SESSION_ID.PATTERN.toString()}`
      ),
    };
  }

  return { success: true, data: undefined };
}

File: packages/swift/Sources/PhaseAnalytics/Core/Utils/Validator.swift
import Foundation

internal struct Validator {

    static func validateDeviceID(_ deviceID: String) -> Result<Void, PhaseError> {
        guard deviceID.count >= ValidationConstants.DeviceID.minLength else {
            return .failure(
                .validationFailed(
                    "Device ID must be at least \(ValidationConstants.DeviceID.minLength) characters"
                ))
        }

        guard deviceID.count <= ValidationConstants.DeviceID.maxLength else {
            return .failure(
                .validationFailed(
                    "Device ID must not exceed \(ValidationConstants.DeviceID.maxLength) characters"
                ))
        }

        guard deviceID.range(of: ValidationConstants.DeviceID.pattern, options: .regularExpression) != nil else {
            return .failure(
                .validationFailed(
                    "Device ID must match pattern: \(ValidationConstants.DeviceID.pattern)"
                ))
        }

        return .success(())
    }

    static func validateSessionID(_ sessionID: String) -> Result<Void, PhaseError> {
        guard sessionID.count >= ValidationConstants.SessionID.minLength else {
            return .failure(
                .validationFailed(
                    "Session ID must be at least \(ValidationConstants.SessionID.minLength) characters"
                ))
        }

        guard sessionID.count <= ValidationConstants.SessionID.maxLength else {
            return .failure(
                .validationFailed(
                    "Session ID must not exceed \(ValidationConstants.SessionID.maxLength) characters"
                ))
        }

        guard sessionID.range(of: ValidationConstants.SessionID.pattern, options: .regularExpression) != nil else {
            return .failure(
                .validationFailed(
                    "Session ID must match pattern: \(ValidationConstants.SessionID.pattern)"
                ))
        }

        return .success(())
    }

    static func validateEventName(_ name: String) -> Result<Void, PhaseError> {
        guard name.count >= ValidationConstants.EventName.minLength else {
            return .failure(
                .validationFailed(
                    "Event name must be at least \(ValidationConstants.EventName.minLength) character(s)"
                ))
        }

        guard name.count <= ValidationConstants.EventName.maxLength else {
            return .failure(
                .validationFailed(
                    "Event name must not exceed \(ValidationConstants.EventName.maxLength) characters"
                ))
        }

        guard name.range(of: ValidationConstants.EventName.pattern, options: .regularExpression) != nil else {
            return .failure(
                .validationFailed(
                    "Event name must match pattern: \(ValidationConstants.EventName.pattern)"
                ))
        }

        return .success(())
    }

    static func validateEventParams(_ params: EventParams?) -> Result<Void, PhaseError> {
        guard let params = params else {
            return .success(())
        }

        let dict = params.dictionary

        let depth = getObjectDepth(dict)
        guard depth <= ValidationConstants.EventParams.maxDepth else {
            return .failure(
                .validationFailed(
                    "Event params exceed maximum depth of \(ValidationConstants.EventParams.maxDepth) (got \(depth))"
                ))
        }

        do {
            let data = try JSONSerialization.data(withJSONObject: dict, options: [])
            guard data.count <= ValidationConstants.EventParams.maxSize else {
                return .failure(
                    .validationFailed(
                        "Event params exceed maximum size of \(ValidationConstants.EventParams.maxSize) bytes"
                    ))
            }
        } catch {
            return .failure(.validationFailed("Event params must be JSON serializable"))
        }

        return .success(())
    }

    private static func getObjectDepth(
        _ obj: Any,
        currentDepth: Int = 1,
        visited: Set<ObjectIdentifier> = []
    ) -> Int {
        if let dict = obj as? [String: Any] {
            let objId = ObjectIdentifier(dict as AnyObject)
            if visited.contains(objId) {
                return currentDepth
            }

            var newVisited = visited
            newVisited.insert(objId)

            var maxDepth = currentDepth
            for value in dict.values {
                let depth = getObjectDepth(value, currentDepth: currentDepth + 1, visited: newVisited)
                maxDepth = max(maxDepth, depth)
            }
            return maxDepth
        }

        if let array = obj as? [Any] {
            let objId = ObjectIdentifier(array as AnyObject)
            if visited.contains(objId) {
                return currentDepth
            }

            var newVisited = visited
            newVisited.insert(objId)

            var maxDepth = currentDepth
            for item in array {
                let depth = getObjectDepth(item, currentDepth: currentDepth + 1, visited: newVisited)
                maxDepth = max(maxDepth, depth)
            }
            return maxDepth
        }

        return currentDepth
    }
}
